gh-103861: Fix Zip64 extensions not being properly applied in some cases

pR0Ps opened the pull request:
This commit fixes an issue where adding a small file to a `ZipFile` object while forcing zip64 extensions causes an extra Zip64 record to be added to the zip, but doesn't update the `min_version` or file sizes.

To create a file that reproduces the issue (copied from #103861):
```python
import zipfile

with zipfile.ZipFile("out.zip", mode="w", allowZip64=True) as zf:
    with zf.open("text.txt", mode="w", force_zip64=True) as zi:
        zi.write(b"some data")
```

Diff of information extracted by [zipdetails](https://github.com/pmqs/zipdetails) from running the above script before and after this commit.
```patch
 0000 LOCAL HEADER #1       04034B50
-0004 Extract Zip Spec      14 '2.0'
+0004 Extract Zip Spec      2D '4.5'
 0005 Extract OS            00 'MS-DOS'
 0006 General Purpose Flag  0000
 0008 Compression Method    0000 'Stored'
 000A Last Mod Time         00210000 'Mon Dec 31 19:00:00 1979'
 000E CRC                   D9C2E91E
-0012 Compressed Length     00000009
-0016 Uncompressed Length   00000009
+0012 Compressed Length     FFFFFFFF
+0016 Uncompressed Length   FFFFFFFF
 001A Filename Length       0008
 001C Extra Length          0014
 001E Filename              'text.txt'
 0026 Extra ID #0001        0001 'ZIP64'
 0028   Length              0010
 002A   Uncompressed Size   0000000000000009
 0032   Compressed Size     0000000000000009
 003A PAYLOAD               some data

 0043 CENTRAL HEADER #1     02014B50
-0047 Created Zip Spec      14 '2.0'
+0047 Created Zip Spec      2D '4.5'
 0048 Created OS            03 'Unix'
-0049 Extract Zip Spec      14 '2.0'
+0049 Extract Zip Spec      2D '4.5'
 004A Extract OS            00 'MS-DOS'
 004B General Purpose Flag  0000
 004D Compression Method    0000 'Stored'
 004F Last Mod Time         00210000 'Mon Dec 31 19:00:00 1979'
 0053 CRC                   D9C2E91E
 0057 Compressed Length     00000009
 005B Uncompressed Length   00000009
 005F Filename Length       0008
 0061 Extra Length          0000
 0063 Comment Length        0000
 0065 Disk Start            0000
 0067 Int File Attributes   0000
      [Bit 0]               0 'Binary Data'
 0069 Ext File Attributes   01800000
 006D Local Header Offset   00000000
 0071 Filename              'text.txt'

 0079 END CENTRAL HEADER    06054B50
 007D Number of this disk   0000
 007F Central Dir Disk no   0000
 0081 Entries in this disk  0001
 0083 Total Entries         0001
 0085 Size of Central Dir   00000036
 0089 Offset to Central Dir 00000043
 008D Comment Length        0000
```

A test has also been added that checks that these properties are correctly set.


<!-- gh-issue-number: gh-103861 -->
* Issue: gh-103861
<!-- /gh-issue-number -->

Potential reviewer based on the git blame of the changed lines: @serhiy-storchaka (182d7cd5)
Potential reviewers based on the [experts index](https://devguide.python.org/core-developers/experts/): @Yhg1s, @gpshead

pR0Ps committed:
Fix Zip64 extensions not being properly applied in some cases

This commit fixes an issue where adding a small file to a `ZipFile`
object while forcing zip64 extensions causes an extra Zip64 record to be
added to the zip, but doesn't update the `min_version` or file sizes.

Fixes #103861

pR0Ps committed:
Fix edge cases in checking if zip64 extensions are required

This fixes an issue where if data requiring zip64 extensions was added
to an unseekable stream without specifying `force_zip64=True`, zip64
extensions would not be used and a RuntimeError would not be raised when
closing the file (even though the size would be known at that point).
This would result in successfully writing corrupt zip files.

Deciding if zip64 extensions are required outside of the `FileHeader`
function means that both `FileHeader` and `_ZipWriteFile` will always be
in sync. Previously, the `FileHeader` function could enable zip64
extensions without propagating that decision to the `_ZipWriteFile`
class, which would then not correctly write the data descriptor record
or check for errors on close.

gpshead commented on Lib/test/test_zipfile/test_core.py:
Can you add @bigmemtest decorators to the tests that actually create and use zipfile.ZIP64_LIMIT byte strings?  See `Lib/test/support/__init__.py` for bigmemtest's definition.

pR0Ps replied:
The `zipfile.ZIP64_LIMIT` is patched down to 1000 in the `AbstractTestZip64InSmallFiles.setUp()` function before running these tests so none of them actually create huge bytestrings.

gpshead commented on Lib/test/test_zipfile/test_core.py:
```diff
@@ -1115,6 +1115,92 @@ def test_force_zip64(self):
+            header, vers, os, flags, comp, csize, usize, fn_len,
+            ex_total_len, filename, ex_id, ex_len, ex_usize, ex_csize,
+            cd_sig
+        ) = struct.unpack("<4sBBHH8xIIHH8shhQQ{}x4s".format(file_size), seekable_data[:62 + file_size])
```
It is unusual to see this kind of complex struct unpack construct logic in a test.

There is no way anyone reading this can say that this struct definition is any more correct than the struct definitions used by the code generating the data it is testing. They'll have to go do research.

At a minimum, include a comment linking back to the spec this lines up with on each of these.  I get the point of this in tests, I just want it to be easier to directly reason about.

pR0Ps replied:
Added links to the spec and referenced the relevant sections - let me know if this isn't what you had in mind.

I agree that it's a bit hard to understand, but since most zip extractors/parsers (including Python's one) are fairly lenient in what they accept, I wanted to make sure that everything was following spec at a byte-level as well as at a functional level.

gpshead commented on Lib/zipfile/__init__.py:
```diff
@@ -439,7 +439,7 @@ def __repr__(self):
         return ''.join(result)
 
-    def FileHeader(self, zip64=None):
+    def FileHeader(self, zip64):
```
We can't change the public API of `ZipInfo.FileHeader()` by making the parameter required without a deprecation period as existing code must be assumed to depend on being able to call it with no parameters.

pR0Ps replied:
Sorry, I figured it was ok to change it since it's not really documented anywhere and seems like a very internal-only method, despite the lack of an `_`-prefix. I've added back the default (but changed it to `False` since it doesn't dynamically pick if it uses zip64 extension or not anymore).

gpshead commented on Lib/test/test_zipfile/test_core.py:
```diff
@@ -1080,6 +1080,127 @@ def test_generated_valid_zip64_extra(self):
+        ) = struct.unpack("<4sBBHH8xIIHH8shhQQx4s", zipdata[:63])
+
+        self.assertEqual(header, b"PK\x03\x04")  # local file header
+        self.assertGreaterEqual(vers, zipfile.ZIP64_VERSION)  # requires zip64 to extract
```
add comments describing where the #103861 bugs were before this test was added to the test.

pR0Ps replied:
Added a comment to `test_force_zip64` mentioning the issue and specifically calling out the bits of data that were wrong. `test_unseekable_zip_known_filesize` also tests this issue but I didn't mention it in the comment on it directly since it's more about the zip64 interaction with the data descriptor than just checking if zip64 extensions are properly used.

pR0Ps committed:
SQUASHME: Add zip64 default back to FileHeader

pR0Ps committed:
SQUASHME: Added comments to tests

gpshead committed:
Restore ZipInfo.FileHeader zip64=None default to maintain backwards compatibility in the API.

Code within this module always passes an explicit zip64, so the overall bug fix remains valid. We just don't want to break existing user code constructing their own ZipInfo objects and calling zi.FileHeader themselves for whatever reasons.  _(hopefully rare, but it isn't a protected or private API so we can't make assumptions)_

gpshead committed:
Merge branch 'main' into bugfix/force-zip64
